src/dotnet/ToolPackage/ToolPackageObtainer.cs

```diff
@@ -42,6 +41,79 @@ IProjectRestorer projectRestorer
             string targetframework = null,
             string source = null,
             string verbosity = null)
+        {
+            var stageDirectory = _toolsPath.WithSubDirectories(".stage", Path.GetRandomFileName());
+
+            var resourceManager = new TransactionResourceManager(
+                commit: () =>
+                {
+                    Directory.Move(
+                        stageDirectory.WithSubDirectories(packageId).Value,
+                        _toolsPath.WithSubDirectories(packageId).Value);
+
+                    Directory.Delete(stageDirectory.Value, true);
+                },
+                prepare: preparingEnlistment =>
+                {
+                    if (Directory.Exists(_toolsPath.WithSubDirectories(packageId).Value))
+                    {
+                        preparingEnlistment.ForceRollback();
```

Comment:
From my experimentation, this call isn't necessary; the transaction manager will rollback all enlistments from a throwing `Prepare` call.

Reply:
Let me try and see if my test passes

Reply:
Actually, I am wrong here.  I was testing my gist code which does handle this correctly; it is not inherent in the transaction manager logic.  If the `Prepare` call fails, we should call `ForceRollback` with an exception that will be raised as an inner exception to `TransactionAbortedException` by the transaction manager after the rollback completes.  The enlistment that aborts the transaction will not be rolled back, so it needs to be invoked manually.

This is what I had in the gist:

```c#
try
{
    Prepare();
}
catch (Exception ex)
{
    Undo();
    preparingEnlistment.ForceRollback(ex);
    return;
}

preparingEnlistment.Prepared();
```

Reply:
should we keep throwing it? instead of retrun

Reply:
With my experimentation, it appears that the transaction manager doesn't expect `Prepare` to throw; instead it expects a call to `ForceRollback` (without giving it an exception, it'll throw a `TransactionAbortedException` with no inner exception).  If `Prepare` throws the transaction is not rolled back.

Reply:
Rollback is empty now